File: DSA/String/Code/ReverseAString.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

// Reverse a String in C++
// Different methods to reverse a string

int main() {
    string str = "Hello World";
    
    // Method 1: Using reverse() algorithm
    cout << "Method 1: Using reverse() algorithm\n";
    string str1 = str;
    cout << "Original: " << str1 << "\n";
    reverse(str1.begin(), str1.end());
    cout << "Reversed: " << str1 << "\n";
    
    // Method 2: Using loop with swap
    cout << "\nMethod 2: Using loop\n";
    string str2 = str;
    cout << "Original: " << str2 << "\n";
    int len = str2.length();
    for (int i = 0; i < len / 2; i++) {
        swap(str2[i], str2[len - 1 - i]);
    }
    cout << "Reversed: " << str2 << "\n";
    
    // Method 3: Building new string
    cout << "\nMethod 3: Building new string\n";
    string str3 = "Programming";
    cout << "Original: " << str3 << "\n";
    string reversed = "";
    for (int i = str3.length() - 1; i >= 0; i--) {
        reversed += str3[i];
    }
    cout << "Reversed: " << reversed << "\n";
    
    // Method 4: Using iterators
    cout << "\nMethod 4: Using reverse iterators\n";
    string str4 = "Iteration";
    cout << "Original: " << str4 << "\n";
    cout << "Reversed: ";
    for (auto it = str4.rbegin(); it != str4.rend(); ++it) {
        cout << *it;
    }
    cout << "\n";
    
    // Method 5: Using recursion
    cout << "\nMethod 5: Using recursion\n";
    string str5 = "Recursion";
    cout << "Original: " << str5 << "\n";
    cout << "Reversed: ";
    
    function<void(const string&, int)> reverseRec = 
        [&](const string& s, int n) {
            if (n >= 0) {
                cout << s[n];
                reverseRec(s, n - 1);
            }
        };
    reverseRec(str5, str5.length() - 1);
    cout << "\n";
    
    return 0;
}

File: DSA/String/Code/ReverseAString.py
# Reverse a String in Python
# Different methods to reverse a string

str1 = "Hello World"

# Method 1: Using slicing
print("Method 1: Using slicing")
original = str1
reversed_str = original[::-1]
print(f"Original: {original}")
print(f"Reversed: {reversed_str}")

# Method 2: Using reversed() function
print("\nMethod 2: Using reversed() function")
original = str1
reversed_str = ''.join(reversed(original))
print(f"Original: {original}")
print(f"Reversed: {reversed_str}")

# Method 3: Using loop
print("\nMethod 3: Using loop")
original = "Programming"
reversed_str = ""
for char in original:
    reversed_str = char + reversed_str
print(f"Original: {original}")
print(f"Reversed: {reversed_str}")

# Method 4: Using list and reverse()
print("\nMethod 4: Using list.reverse()")
original = "Python"
chars = list(original)
chars.reverse()
reversed_str = ''.join(chars)
print(f"Original: {original}")
print(f"Reversed: {reversed_str}")

# Method 5: Using recursion
print("\nMethod 5: Using recursion")
def reverse_recursive(s):
    if len(s) == 0:
        return s
    else:
        return reverse_recursive(s[1:]) + s[0]

original = "Recursion"
reversed_str = reverse_recursive(original)
print(f"Original: {original}")
print(f"Reversed: {reversed_str}")

# Method 6: Using reduce
print("\nMethod 6: Using reduce")
from functools import reduce
original = "Reduce"
reversed_str = reduce(lambda x, y: y + x, original)
print(f"Original: {original}")
print(f"Reversed: {reversed_str}")

# Method 7: Using stack concept
print("\nMethod 7: Using stack")
original = "Stack"
stack = list(original)
reversed_str = ""
while stack:
    reversed_str += stack.pop()
print(f"Original: {original}")
print(f"Reversed: {reversed_str}")
